core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/messaginglayer/FlowControl.scala

```diff
@@ -0,0 +1,38 @@
+package edu.uci.ics.amber.engine.architecture.messaginglayer
+
+import akka.actor.Cancellable
+import edu.uci.ics.amber.engine.common.virtualidentity.ActorVirtualIdentity
+
+import scala.collection.mutable
+
+/**
+  * We implement credit-based flow control. Suppose a sender worker S sends data in batches to a receiving worker R
+  * using the network communicator actor NC. The different parts of flow control work as follows:
+  *
+  * 1. A worker has a fixed amount of credits for each of its sender workers. When R is expensive, its internal queue
+  * starts getting filled with data from S. This leads to a decrease in credits available for S.
+  *
+  * 2. R sends the credits available in the NetworkAck() being sent to S. This includes acks for data messages and
+  * control messages. The responsibility to decrease data sending lies on S now.
+  *
+  * 3. Upon receiving NetworkAck(), S saves the credit information and then does two things:
+  * a) Tell parent to enable/disable backpressure: If the `buffer in NC` + `credit available in R` is less
+  * than the `backlog of data in NC`, backpressure needs to be enabled. For this, the NC sends a control message
+  * to S to enable backpressure (pause data processing) and adds R to `overloaded` list. On the other hand, if
+  * `buffer in NC` + `credit available in R` is enough, then R is removed from the overloaded list (if present). If
+  * the `overloaded` list is empty, then NC sends a request to S to disable backpressure (resume processing).
+  *
+  * b) It looks at its backlog and sends an amount of data, less than credits available, to R.
+  *
+  * 3. If R sends a credit of 0, then S won't send any data as a response to NetworkAck(). This will lead to a problem
+  * because then there is no way for S to know when the data in its congestion control module can be sent. Thus,
+  * whenever S receives a credit of 0, it registers a periodic callback that serves as a trigger for it to send
+  * credit poll request to R. Then, R responds with a NetworkAck() for the credits.
+  */
+class FlowControl {
+  val receiverIdToDataBacklog = new mutable.HashMap[ActorVirtualIdentity, Int]()
```

Comment:
Can we remove some of these data structures? I see some of them are redundant. For example, for the backlog map, can we just check the buffer size somewhere? For the overloaded receivers, can we just look at the credit map?

Reply:
we can replace with functions, for example overloaded receivers

Reply:
we can replace with functions, for example overloaded receivers

Reply:
done